diff --git a/core/writer.c b/core/writer.c
index 54919868..2f4838fb 100644
--- a/core/writer.c
+++ b/core/writer.c
@@ -529,7 +529,7 @@ sendbody:
                 	break;
                 }
                 if (!uwsgi_is_again()) continue;
-                ret = uwsgi_wait_write_req(wsgi_req);
+                ret = uwsgi.wait_write_hook(wsgi_req->fd, -1);
                 if (ret < 0) { wsgi_req->write_errors++; return -1;}
                 if (ret == 0) {
                 	uwsgi_log("uwsgi_response_writev_body_do() TIMEOUT !!!\n");
