diff --git a/Makefile b/Makefile
index a7d6d3c92..45377fb17 100644
--- a/Makefile
+++ b/Makefile
@@ -254,7 +254,7 @@ define build_proj_tmpl
 PROS += $$(basename $$(value $(1)_PRO)).build
 $(1)_MAKE := $$(basename $$(value $(1)_PRO)).build/Makefile
 $$(value $(1)): $$(value $(1)_MAKE)
-	cd $$(dir $$(value $(1)_MAKE)) && $(MAKE);
+	+$(MAKE) -C $$(dir $$(value $(1)_MAKE))
 endef
 
 .PHONY : all clean deploy
@@ -309,7 +309,7 @@ clean:
 	rm -rf $(TARGETS) $(ARCHIVE)
 	for i in $(PROS); do \
 		if [ -d $$i -a -f $$i/Makefile ]; then \
-			cd $$i && $(MAKE) distclean; \
+			$(MAKE) -C $$i distclean; \
 		fi \
 done
 
