diff --git a/ASCOfficePPTXFile/ASCOfficeDrawingConverter.cpp b/ASCOfficePPTXFile/ASCOfficeDrawingConverter.cpp
index 60610b970..aecf56fe4 100644
--- a/ASCOfficePPTXFile/ASCOfficeDrawingConverter.cpp
+++ b/ASCOfficePPTXFile/ASCOfficeDrawingConverter.cpp
@@ -64,7 +64,7 @@ const double g_emu_koef	= 25.4 * 36000 / 72.0;
 void DUMP_MESSAGE_TO_FILE(const char* strMessage)
 {
 	FILE* file = fopen("c:\\1.txt", "a+");
-	fprintf(file, strMessage);
+	fputs(strMessage, file);
 	fclose(file);
 }
 
diff --git a/DesktopEditor/doctrenderer/nativecontrol.h b/DesktopEditor/doctrenderer/nativecontrol.h
index af12384c9..9a4ef6d1d 100644
--- a/DesktopEditor/doctrenderer/nativecontrol.h
+++ b/DesktopEditor/doctrenderer/nativecontrol.h
@@ -375,9 +375,9 @@ public:
             FILE* _file = NSFile::CFileBinary::OpenFileNative(m_sChangesBuilderPath, L"a+");
             if (NULL != _file)
             {
-                fprintf(_file, "\"");
-                fprintf(_file, sParam.c_str());
-                fprintf(_file, "\",");
+                fputs("\"", _file);
+                fputs(sParam.c_str(), _file);
+                fputs("\",", _file);
                 fclose(_file);
             }
         }
diff --git a/DesktopEditor/fontengine/application_generate_fonts_common.h b/DesktopEditor/fontengine/application_generate_fonts_common.h
index f4bcb82ea..263e84d04 100644
--- a/DesktopEditor/fontengine/application_generate_fonts_common.h
+++ b/DesktopEditor/fontengine/application_generate_fonts_common.h
@@ -106,8 +106,8 @@ namespace NSCommon
     {
         FILE* f = fopen(sLogFile.c_str(), "a+");
         std::string sDataA = U_TO_UTF8(sData);
-        fprintf(f, sDataA.c_str());
-        fprintf(f, "\n");
+        fputs(sDataA.c_str(), f);
+        fputs("\n", f);
         fclose(f);
     }
 }
