diff --git a/DesktopEditor/AllFontsGen/main.cpp b/DesktopEditor/AllFontsGen/main.cpp
index 97a16c816..fea10ec08 100644
--- a/DesktopEditor/AllFontsGen/main.cpp
+++ b/DesktopEditor/AllFontsGen/main.cpp
@@ -297,7 +297,7 @@ namespace NSCommon
 
         // теперь строим массив всех шрифтов по имени
         std::map<std::wstring, CFontInfoJS> mapFonts;
-        CArray<std::wstring> arrFonts;
+        std::vector<std::wstring> arrFonts;
 
         for (std::vector<NSFonts::CFontInfo*>::iterator i = pList->begin(); i != pList->end(); i++)
         {
@@ -368,25 +368,14 @@ namespace NSCommon
                 }
 
                 mapFonts.insert(std::pair<std::wstring, CFontInfoJS>(fontInfo.m_sName, fontInfo));
-                arrFonts.Add(fontInfo.m_sName);
+                arrFonts.push_back(fontInfo.m_sName);
             }
         }
         // -------------------------------------------
 
-        // теперь сортируем шрифты по имени ----------
-        int nCountFonts = arrFonts.GetCount();
-        for (int i = 0; i < nCountFonts; ++i)
-        {
-            for (int j = i + 1; j < nCountFonts; ++j)
-            {
-                if (arrFonts[i] > arrFonts[j])
-                {
-                    std::wstring temp = arrFonts[i];
-                    arrFonts[i] = arrFonts[j];
-                    arrFonts[j] = temp;
-                }
-            }
-        }
+        int nCountFonts = arrFonts.size();
+        std::sort(arrFonts.begin(), arrFonts.end());
+
         // -------------------------------------------
         if (L"" != strFolderThumbnails)
         {
@@ -736,7 +725,7 @@ namespace NSCommon
                     }
 
                     int nSize = 0;
-                    if (-1 != info.m_lIndexR)
+                    if (-1 != info.m_lIndexR && info.m_lIndexR < mapFontFiles2.size())
                     {
                         NSFile::CFileBinary oFile;
                         if (oFile.OpenFile(mapFontFiles2[info.m_lIndexR]))
@@ -746,7 +735,7 @@ namespace NSCommon
                                 nSize = nTmp;
                         }
                     }
-                    if (-1 != info.m_lIndexB)
+                    if (-1 != info.m_lIndexB && info.m_lIndexB < mapFontFiles2.size())
                     {
                         NSFile::CFileBinary oFile;
                         if (oFile.OpenFile(mapFontFiles2[info.m_lIndexB]))
@@ -756,7 +745,7 @@ namespace NSCommon
                                 nSize = nTmp;
                         }
                     }
-                    if (-1 != info.m_lIndexI)
+                    if (-1 != info.m_lIndexI && info.m_lIndexI < mapFontFiles2.size())
                     {
                         NSFile::CFileBinary oFile;
                         if (oFile.OpenFile(mapFontFiles2[info.m_lIndexI]))
@@ -766,7 +755,7 @@ namespace NSCommon
                                 nSize = nTmp;
                         }
                     }
-                    if (-1 != info.m_lIndexBI)
+                    if (-1 != info.m_lIndexBI && info.m_lIndexBI < mapFontFiles2.size())
                     {
                         NSFile::CFileBinary oFile;
                         if (oFile.OpenFile(mapFontFiles2[info.m_lIndexBI]))
