diff --git a/src/apps/relay/mainrelay.c b/src/apps/relay/mainrelay.c
index 79e13b9..e5e30be 100644
--- a/src/apps/relay/mainrelay.c
+++ b/src/apps/relay/mainrelay.c
@@ -1429,6 +1429,8 @@ static void set_option(int c, char *value)
 		use_lt_credentials = 1;
 		break;
 	case STATIC_AUTH_SECRET_VAL_OPT:
+		if (value[0] == '!' && value[1] == '\0')
+			value = getenv("COTURN_STATIC_AUTH_SECRET");
 		add_to_secrets_list(&turn_params.default_users_db.ram_db.static_auth_secrets,value);
 		turn_params.use_auth_secret_with_timestamp = 1;
         use_tltc = 1;
