diff --git a/lib/Config.php b/lib/Config.php
index 6ce476c7..a600158f 100644
--- a/lib/Config.php
+++ b/lib/Config.php
@@ -216,6 +216,9 @@ class Config {
 			$server = $servers[0];
 		}
 
+		if ($server['secret'] === '__FROM_ENV')
+			$server['secret'] = $_ENV['COTURN_STATIC_AUTH_SECRET'];
+
 		// Credentials are valid for 24h
 		// FIXME add the TTL to the response and properly reconnect then
 		$timestamp = $this->timeFactory->getTime() + 86400;
