diff --git a/rel/files/nodetool b/rel/files/nodetool
index 6cb027336..ddc7cbdac 100755
--- a/rel/files/nodetool
+++ b/rel/files/nodetool
@@ -1,6 +1,6 @@
 %% -*- mode: erlang;erlang-indent-level: 4;indent-tabs-mode: nil -*-
 %% ex: ft=erlang ts=4 sw=4 et
-%%! -args_file {{mongooseim_nodetool_etc_dir}}/vm.dist.args
+%%! -args_file {{mongooseim_nodetool_etc_dir}}/vm.dist.args -setcookie none
 %% -------------------------------------------------------------------
 %%
 %% nodetool: Helper Script for interacting with live nodes
