diff --git a/rel/files/mongooseim b/rel/files/mongooseim
index a4b8dbfb8..3971b071b 100755
--- a/rel/files/mongooseim
+++ b/rel/files/mongooseim
@@ -12,9 +12,12 @@ PIPE_DIR=/tmp/mongooseim_pipe_`whoami`/
 RUNNER_USER={{mongooseim_runner_user}}
 
 EJABBERD_SO_PATH=`ls -dt "$RUNNER_BASE_DIR"/lib/mongooseim-*/priv/lib | head -1`
-EJABBERD_CONFIG_PATH="$RUNNER_ETC_DIR"/mongooseim.cfg
 export EJABBERD_SO_PATH
-export EJABBERD_CONFIG_PATH
+
+if [ -z "$EJABBERD_CONFIG_PATH" ]; then
+    EJABBERD_CONFIG_PATH="$RUNNER_ETC_DIR"/mongooseim.cfg
+    export EJABBERD_CONFIG_PATH
+fi
 
 # Make sure this script is running as the appropriate user
 if [ ! -z "$RUNNER_USER" ] && [ `whoami` != "$RUNNER_USER" ]; then
