diff --git a/configure.ac b/configure.ac
index 2f05fd972..5692f4c5d 100644
--- a/configure.ac
+++ b/configure.ac
@@ -45,11 +45,6 @@ AC_PROG_CC
 
 AC_LANG_PUSH([C++])
 
-AS_IF([test `uname -s` = Linux],
-      [AC_PATH_PROG([SETCAP], setcap,[],[/bin:/usr/bin:/sbin:/usr/sbin])]
-       AS_IF([test -z "$SETCAP"],
-       [AC_MSG_ERROR([The setcap command is needed. It might be in a package called libcap-progs.])]))
-
 # Declare options
 AC_ARG_ENABLE([debug],
               AS_HELP_STRING([--enable-debug],
